test/infrastructure/docker/internal/controllers/dockermachine_controller.go

```diff
@@ -267,7 +268,17 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, cluster *
 	// we should only do this once, as reconfiguration more or less ensures
 	// node ref setting fails
 	if util.IsControlPlaneMachine(machine) && !dockerMachine.Status.LoadBalancerConfigured {
-		if err := externalLoadBalancer.UpdateConfiguration(ctx); err != nil {
+		var unsafeLoadBalancerConfigCM *corev1.ConfigMap
+		if dockerCluster.Spec.LoadBalancer.UnsafeHAProxyConfigTemplateRef != nil {
+			unsafeLoadBalancerConfigCM = &corev1.ConfigMap{}
+			if err := r.Get(ctx, types.NamespacedName{
+				Name:      dockerCluster.Spec.LoadBalancer.UnsafeHAProxyConfigTemplateRef.Name,
+				Namespace: dockerMachine.Namespace,
+			}, unsafeLoadBalancerConfigCM); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to retrieve custom HAProxy configuration ConfigMap")
+			}
+		}
+		if err := externalLoadBalancer.UpdateConfiguration(ctx, unsafeLoadBalancerConfigCM); err != nil {
```

Comment:
Because this code is duplicated below: what about creating a helper function and using that one instead? (note, I paired the suggestion together with passing just a string to the `UpdateConfiguration` function)

```suggestion
		unsafeLoadBalancerConfigTemplate, err := getUnsafeLoadBalancerConfigTemplate(ctx)
		if err != nil {
				return ctrl.Result{}, errors.Wrap(err, "failed to retrieve HAProxy configuration from UnsafeHAProxyConfigTemplateRef")
		}
		if err := externalLoadBalancer.UpdateConfiguration(ctx, unsafeLoadBalancerConfigTemplate); err != nil {
```

```go
func (r *DockerMachineReconciler) getUnsafeLoadBalancerConfigTemplate (ctx context.Context, dockerCluster *infrav1.DockerCluster) (string, error) {
	if dockerCluster.Spec.LoadBalancer.UnsafeHAProxyConfigTemplateRef == nil {
		return "", nil
	}
	var cm *corev1.ConfigMap
	key := types.NamespacedName{
		Name:      dockerCluster.Spec.LoadBalancer.UnsafeHAProxyConfigTemplateRef.Name,
		Namespace: dockerMachine.Namespace,
	}
	if err := r.Get(ctx, key, cm ); err != nil {
		return "", errors.Wrapf(err, "failed to retrieve custom HAProxy configuration ConfigMap %s", key)
	}
	template, ok := unsafeLoadBalancerConfigCM.Data["value"]
	if !ok {
		return fmt.Errorf("Expected key \"value\" to exist in ConfigMap %s", key)
	}
	return template, nil
}
```

This also allows to return an error if the key does not exist.